pkg/oam/workload/translate.go

```diff
@@ -0,0 +1,136 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	workloadv1alpha1 "github.com/crossplane/crossplane/apis/workload/v1alpha1"
+)
+
+const (
+	errWrapInKubeApp = "unable to wrap objects in KubernetesApplication"
+)
+
+var (
+	serviceKind       = reflect.TypeOf(corev1.Service{}).Name()
+	serviceAPIVersion = corev1.SchemeGroupVersion.String()
+)
+
+// LabelKey is the label applied to translated workload objects.
+const LabelKey = "workload.oam.crossplane.io"
+
+// KubeAppWrapper wraps a set of translated objects in a KubernetesApplication.
+func KubeAppWrapper(ctx context.Context, w resource.Workload, objs []resource.Object) ([]resource.Object, error) {
+	if objs == nil {
+		return nil, nil
+	}
+
+	app := &workloadv1alpha1.KubernetesApplication{}
+
+	for _, o := range objs {
+		b, err := json.Marshal(o)
+		if err != nil {
+			return nil, errors.Wrap(err, errWrapInKubeApp)
+		}
+
+		kart := workloadv1alpha1.KubernetesApplicationResourceTemplate{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf("%s-%s", o.GetName(), strings.ToLower(o.GetObjectKind().GroupVersionKind().Kind)),
+				Labels: map[string]string{
+					LabelKey: string(w.GetUID()),
+				},
+			},
+			Spec: workloadv1alpha1.KubernetesApplicationResourceSpec{
+				Template: runtime.RawExtension{Raw: b},
+			},
+		}
+
+		app.Spec.ResourceTemplates = append(app.Spec.ResourceTemplates, kart)
+	}
+
+	app.SetName(w.GetName())
+
+	app.Spec.ResourceSelector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			LabelKey: string(w.GetUID()),
+		},
+	}
+
+	return []resource.Object{app}, nil
+}
+
+// ServiceInjector adds a Service object for the first Port on the first
+// Container for the first Deployment observed in a workload translation.
+func ServiceInjector(ctx context.Context, w resource.Workload, objs []resource.Object) ([]resource.Object, error) {
+	if objs == nil {
+		return nil, nil
+	}
+
+	for _, o := range objs {
+		d, ok := o.(*appsv1.Deployment)
+		if !ok {
+			continue
+		}
+
+		if len(d.Spec.Template.Spec.Containers) > 0 {
+			s := &corev1.Service{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       serviceKind,
+					APIVersion: serviceAPIVersion,
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: d.GetName(),
+					Labels: map[string]string{
+						LabelKey: string(w.GetUID()),
+					},
+				},
+				Spec: corev1.ServiceSpec{
+					Selector: d.Spec.Selector.MatchLabels,
+					Ports:    []corev1.ServicePort{},
+					Type:     corev1.ServiceTypeLoadBalancer,
+				},
+			}
+
+			if len(d.Spec.Template.Spec.Containers[0].Ports) > 0 {
+				s.Spec.Ports = []corev1.ServicePort{
+					{
+						Name:       d.GetName(),
+						Port:       d.Spec.Template.Spec.Containers[0].Ports[0].ContainerPort,
+						TargetPort: intstr.FromInt(int(d.Spec.Template.Spec.Containers[0].Ports[0].ContainerPort)),
+					},
+				}
+			}
+			objs = append(objs, s)
+			break
```

Comment:
This `ServiceInjector` receives a slice of objects and appends to it. So in the case of the local `ContainerizedWorkload`, it will receive a slice with one `Deployment` object and append a `Service` to the slice, then both objects will be created :) It is probably easier to see this here: https://github.com/crossplane/crossplane-runtime/blob/da385a76bc9b8bb68ca2d5685c8d13009c437481/pkg/reconciler/oam/workload/reconciler.go#L159